formiodata/components/grid_base.py

```diff
@@ -107,6 +107,18 @@ def child_component_owner(self):
     def initEmpty(self):
         return self.raw.get('initEmpty')
 
+    def validation_errors(self):
+        errors = []
+        for row_idx, row in enumerate(self.rows):
+            for component_key, component in row.input_components.items():
+                component_errors = component.validation_errors()
+                if bool(component_errors):
+                    # scalar (not grid) components get a Dict from
+                    # method validation_errors()
+                    vals = {component_key: component_errors}
+                    errors.append(vals)
```

Comment:
This looks incorrect. I think there's a level missing. Try writing a test where one grid row contains multiple components, which all have errors. I think now this will look like the first component ends up on line 0 and the second component on line 1.

Speaking of which, shouldn't `errors` be a dict here, so that you can use `row_idx` to indicate the correct row? You probably also need a test with a component on a grid where the first line has no errors and the second line does. If I read this code correctly, it will appear as if the error exists on row 0.

Reply:
Alternatively, instead of having `errors` be a dict, you could simply unconditionally append `vals` to `errors`. That means you'd get a row for each row in the input, but with empty dicts if there are no errors on those rows.

Reply:
@sjamaan Glad if you can review after this 2nd commit.

I tried to fix the `baseGridComponent` its `validtion_errors()` method.
Also extended the unittest, which also points out the following case.

In order to keep the rows indexing similar as the Form (submission) data, I needed to add an extra object/value to represent a valid row.
Therefor I just added the `False` object/value, as show in the example below.
Any second thoughts, is this wrong? Or would `None` be a better value (annotation) here?

```python
'dataGrid': [
    {'textField': {'required': 'Text Field is required'}},
    False, 
    {'textField': {'required': 'Text Field is required'}},
    False
]
```

Reply:
> In order to keep the rows indexing similar as the Form (submission) data, I needed to add an extra object/value to represent a valid row. Therefor I just added the `False` object/value, as show in the example below. Any second thoughts, is this wrong? Or would `None` be a better value (annotation) here?

Like I pointed out in my original message, I think `{}` would actually be better here, as it would potentially reduce the special casing required to traverse the errors.

For convenience, you could have an `ensure_valid()` method which raises an exception with the validation errors if there are any, otherwise just returns (`None` or `True`, whatever is convenient). This would really simplify the code required to handle it. The validation errors could just be wrapped in a custom exception class which only provides an accessor to extract the value.

Alternatively, perhaps a predicate `is_valid()` which returns `True` or `False` could be a nice convenience wrapper.

Reply:
Note that the main `validation_errors` method already also returns `{}` when there are no errors. So having `{}` in the rows would be consistent with the form itself.

Reply:
> Note that the main `validation_errors` method already also returns `{}` when there are no errors. So having `{}` in the rows would be consistent with the form itself.

I want to keep the datastructure retrieved form the  `validation_errors()` method similar with the a form's (JSON) submission which is also a dictionary.

For the rows I still think a list `[ ]` would be the best choice, because a value of grid component is also stored as a list `[ ]` inside the main dictionary (the form submission/JSON).
**So why instead use a dictionary for a grid component it's rows ?**
Which is rather inconsitent, inconvenient and not possible IMHO.

Reply:
The helper methods (exception, is_valid) are a nice idea. Thanks!

Reply:
> > Note that the main `validation_errors` method already also returns `{}` when there are no errors. So having `{}` in the rows would be consistent with the form itself.
> 
> I want to keep the datastructure retrieved form the `validation_errors()` method similar with the a form's (JSON) submission which is also a dictionary.
> 
> For the rows I still think a list `[ ]` would be the best choice, because a value of grid component is also stored as a list `[ ]` inside the main dictionary (the form submission/JSON). **So why instead use a dictionary for a grid component it's rows ?** Which is rather inconsitent, inconvenient and not possible IMHO.

You're already using a dictionary for the row contents. That's what I'm saying - use `{}` instead of `None` when there are no errors on a given row. The rows themselves can be stored in a list, as you do now.

Reply:
@sjamaan Do you agree with the following change? https://github.com/novacode-nl/python-formio-data/pull/40/commits/d27b3fdbdf58bcd35804a18d92b9f917962a6cc5

I would like to make a release, or still work left you think?

Reply:
Looks fine now! Feel free to release.